### PR TITLE
Add: Client setting gui.start_spectator

### PR DESCRIPTION
**Pull request by SamuXarick**
Activates a spectator slot in single player and initiate games as a spectator

### CODE REVIEW

File: src/cheat_gui.cpp [hunk 1]
```diff
@@ -67,8 +67,8 @@ static int32 ClickMoneyCheat(int32 p1, int32 p2)
  */
 static int32 ClickChangeCompanyCheat(int32 p1, int32 p2)
 {
-	while ((uint)p1 < Company::GetPoolSize()) {
-		if (Company::IsValidID((CompanyID)p1)) {
+	while ((uint)p1 <= COMPANY_SPECTATOR) {
```
Comment (glx22) [inline src/cheat_gui.cpp:70]: I think that's wrong, you should still check Company::GetPoolSize(), but you can add something to the test

File: src/company_gui.cpp [hunk 2]
```diff
@@ -2312,13 +2312,13 @@ struct CompanyWindow : Window
 			}
 
 			/* Multiplayer buttons. */
-			plane = ((!_networking) ? (int)SZSP_NONE : (int)(local ? CWP_MP_C_PWD : CWP_MP_C_JOIN));
+			plane = ((!_networking) ? (_settings_client.gui.start_spectator || _local_company == COMPANY_SPECTATOR) ? (_local_company == COMPANY_SPECTATOR) ? (int)CWP_MP_C_JOIN : (int)SZSP_NONE : (int)SZSP_NONE : (int)(local ? CWP_MP_C_PWD : CWP_MP_C_JOIN));
```
Comment (glx22) [inline src/company_gui.cpp]: unreadable :)

File: src/company_gui.cpp [hunk 3]
```diff
@@ -2312,13 +2312,13 @@ struct CompanyWindow : Window
...
 			wi = this->GetWidget<NWidgetStacked>(WID_C_SELECT_MULTIPLAYER);
 			if (plane != wi->shown_plane) {
 				wi->SetDisplayedPlane(plane);
 				reinit = true;
 			}
-			this->SetWidgetDisabledState(WID_C_COMPANY_JOIN,   c->is_ai);
+			this->SetWidgetDisabledState(WID_C_COMPANY_JOIN, c->is_ai || !_networking && _settings_client.gui.start_spectator && local);
```
Comment (glx22) [inline src/company_gui.cpp]: parenthesis please

File: src/company_gui.cpp [hunk 4]
```diff
@@ -2589,28 +2589,38 @@ struct CompanyWindow : Window
...
-					NetworkClientRequestMove(company);
+			case WID_C_COMPANY_JOIN:
+				if (!_networking && (_settings_client.gui.start_spectator || _local_company == COMPANY_SPECTATOR)) {
+					SetLocalCompany((CompanyID)this->window_number);
 				}
 				break;
+		}
+
+#ifdef ENABLE_NETWORK
+		if (_networking) {
```
Comment (glx22) [inline src/company_gui.cpp]: can't you test _networking in the above switch ?

File: src/network/network.cpp [hunk 5]
```diff
@@ -710,7 +710,7 @@ static void NetworkInitGameInfo()
 	/* There should be always space for the server. */
 	assert(NetworkClientInfo::CanAllocateItem());
 	NetworkClientInfo *ci = new NetworkClientInfo(CLIENT_ID_SERVER);
-	ci->client_playas = _network_dedicated ? COMPANY_SPECTATOR : COMPANY_FIRST;
+	ci->client_playas = _network_dedicated || _settings_client.gui.start_spectator ? COMPANY_SPECTATOR : COMPANY_FIRST;
```
Comment (glx22) [inline src/network/network.cpp]: parenthesis

File: src/openttd.cpp [hunk 6]
```diff
@@ -1115,7 +1117,7 @@ void SwitchToMode(SwitchMode new_mode)
 				}
 				/* Update the local company for a loaded game. It is either always
 				 * company #1 (eg 0) or in the case of a dedicated server a spectator */
-				SetLocalCompany(_network_dedicated ? COMPANY_SPECTATOR : COMPANY_FIRST);
+				SetLocalCompany(_network_dedicated || _settings_client.gui.start_spectator ? COMPANY_SPECTATOR : COMPANY_FIRST);
```
Comment (glx22) [inline src/openttd.cpp]: parenthesis

File: src/toolbar_gui.cpp [hunk 7]
```diff
@@ -245,6 +245,20 @@ static void PopupMainCompanyToolbMenu(Window *w, int widget, int grey = 0, bool
...
+			if (widget == WID_TN_COMPANIES) {
+				if (_local_company == COMPANY_SPECTATOR) {
+					*list->Append() = new DropDownListStringItem(STR_NETWORK_COMPANY_LIST_NEW_COMPANY, CTMN_NEW_COMPANY, Company::GetNumItems() >= MAX_COMPANIES);
+				} else {
+					*list->Append() = new DropDownListStringItem(STR_NETWORK_COMPANY_LIST_SPECTATE, CTMN_SPECTATE, false);
+				}
+			} else {
+				*list->Append() = new DropDownListStringItem(STR_NETWORK_TOOLBAR_LIST_SPECTATOR, CTMN_SPECTATOR, false);
+			}
+		}
```
Comment (glx22) [inline src/toolbar_gui.cpp]: that's a lot of copy pasting, but maybe no other way

File: src/company_gui.cpp [hunk 8]
```diff
@@ -2311,8 +2311,21 @@ struct CompanyWindow : Window
...
+				} else {
+					plane = (int)SZSP_NONE;
+				}
+			} else {
+				/* Multiplayer buttons. */
+				if (local) {
+					plane = CWP_MP_C_PWD;
+				} else {
+					plane = CWP_MP_C_JOIN;
+				}
```
Comment (PeterN) [inline src/company_gui.cpp]: Excessive expansion, `plane = local ? CWP_MP_C_PWD : CWP_MP_C_JOIN;` would be fine, and similar for the block above.

File: src/company_gui.cpp [hunk 9]
```diff
@@ -2593,8 +2606,14 @@ struct CompanyWindow : Window
 			case WID_C_COMPANY_PASSWORD:
 				if (this->window_number == _local_company) ShowNetworkCompanyPasswordWindow(this);
 				break;
+#endif /* ENABLE_NETWORK */
 
 			case WID_C_COMPANY_JOIN: {
+				if (!_networking && _local_company == COMPANY_SPECTATOR) {
```
Comment (glx22) [inline src/company_gui.cpp]: If not _networking and _local_company is not spectator you'll execute network specific part. You can probably move the _local_company test just before SetLocalCompany call

File: src/toolbar_gui.cpp [hunk 10]
```diff
@@ -635,6 +646,22 @@ static CallBackFunction MenuClickCompany(int index)
...
+	if (!_networking && _local_company == COMPANY_SPECTATOR) {
+		if (index == CTMN_NEW_COMPANY) {
+			extern Company *DoStartupNewCompany(bool is_ai, CompanyID company = INVALID_COMPANY);
+			Company *c = DoStartupNewCompany(false);
+			c->settings = _settings_client.company;
+			SetLocalCompany(c->index);
+			return CBF_NONE;
+		}
+
+		if (index == CTMN_SPECTATE) {
```
Comment (glx22) [inline src/toolbar_gui.cpp]: spectate button when already spectator ?

**Review by glx22 — COMMENTED** [review]
I think you should not touch economy.cpp and bankruptcy handling

**Comment by TrueBrain** [comment]
I fail to see the usecase for this behavior. Why would I want to be a spectator in a SinglePlayer game? Feels like an anti-pattern. And given the PR does not argue why it should be added, going to close the PR.

From a short chat on IRC, I understand it is to watch the AI do its thing. I don't see how it is useful to change so many lines of code for a very low usercount that will ever use it that way. So if that is the argument, I am still going to close the PR.

Sorry, not every feature can make it into OpenTTD :) But thank you for your contribution nevertheless!

Feel free to rebuttal. But at always, do this with new arguments. Tnx!

**Pull request closed without merging**